Print auth help message based on environment

spenserblack opened the pull request:
<!--
  Thank you for contributing to GitHub CLI!
  To reference an open issue, please write this in your description: `Fixes #NUMBER`
-->
This will check if `gh` is running in a CI environment by checking the `CI` environment variable. If so, it will suggest using environment variables to set the auth token. Additionally, if the CI environment is detected to be GitHub Actions (`GITHUB_ACTION!=""`), it will provide some example code (https://github.com/cli/cli/issues/5387#issuecomment-1150994598).

If it is not a CI environment, it will print the standard help message.

Fixes #5387

spenserblack committed:
Print auth help message based on environment

This will check if `gh` is running in a CI environment. If so, it will
suggest using environment variables to set the auth token. Additionally,
if the CI environment is detected to be GitHub Actions, it will provide
some example code. If it is not a CI environment, it will print the
standard help message.

spenserblack committed:
Check if `GITHUB_ACTION` has value

Instead of checking if the environment variable exists, checks if it has
a value. This makes it easier to test.

spenserblack committed:
Add tests

spenserblack committed:
Use `Fprint` instead of `Fprintf`

Resolves lint error.

mislav commented on cmd/gh/main.go:
```diff
@@ -319,6 +318,39 @@ func printError(out io.Writer, err error, cmd *cobra.Command, debug bool) {
+	var helpMessage string
+	if os.Getenv("CI") == "true" {
+		helpMessage = "Either `GH_TOKEN` or `GITHUB_TOKEN` must be set to use `gh`"
+		if os.Getenv("GITHUB_ACTION") != "" {
```
Please check for GITHUB_ACTIONS (plural) https://docs.github.com/en/actions/learn-github-actions/environment-variables#default-environment-variables
```suggestion 
		if os.Getenv("GITHUB_ACTIONS") == "true" {
```

mislav commented on cmd/gh/main.go:
```diff
@@ -319,6 +318,39 @@ func printError(out io.Writer, err error, cmd *cobra.Command, debug bool) {
+				  GH_TOKEN: ${{ secrets.GITHUB_TOKEN }}
+			`)
+		} else {
+			helpMessage += heredoc.Doc(`
```
The way `helpMessage` is constructed step-by-step eliminates some duplication but makes the whole thing hard to read. I would be more in favor of the duplication if that means that the whole help text can be embedded as a single heredoc block.

mislav replied:
The way `helpMessage` is constructed step-by-step eliminates some duplication but makes the whole thing hard to read. I would be more in favor of the duplication if that means that the whole help text can be embedded as a single heredoc block; no interpolation or joining strings.

mislav replied:
The way `helpMessage` is constructed step-by-step eliminates some duplication but makes the whole thing hard to read. I would be more in favor of the duplication if that means that the whole help text can be embedded as a single heredoc block.

mislav commented on cmd/gh/main.go:
```diff
@@ -319,6 +318,39 @@ func printError(out io.Writer, err error, cmd *cobra.Command, debug bool) {
+func printNoAuthHelp(out io.Writer, cs *iostreams.ColorScheme) {
+	var helpMessage string
+	if os.Getenv("CI") == "true" {
+		helpMessage = "Either `GH_TOKEN` or `GITHUB_TOKEN` must be set to use `gh`"
```
Let's not confuse the user by offering them two competing options that are equivalent. Let's just suggest to set GH_TOKEN

mislav reviewed (changes requested):
Looking good so far!

spenserblack committed:
Use `GITHUB_ACTIONS` environment variable

This environment variable is always set to `true` in a GitHub Action.

Co-authored-by: Mislav Marohnić <hi@mislav.net>

spenserblack committed:
Remove suggestion for setting `GITHUB_TOKEN`

`GITHUB_TOKEN` and `GH_TOKEN` are competing options and may confuse the
user. This changes to only suggesting `GH_TOKEN` if not authorized in a
CI environment.

Co-authored-by: Mislav Marohnić <hi@mislav.net>

spenserblack committed:
Build help message fully in heredoc block

Instead of building the help message step-by-step, constructs the full
help message in a single heredoc function call to make the code more
readable.

spenserblack commented:
@mislav Thanks for the feedback!

mislav committed:
:nail_care:

mislav reviewed (approved):
Thank you! I've additionally simplified a bit.
